mesonbuild/cargo/interpreter.py

```diff
@@ -0,0 +1,449 @@
+
+
+def _convert_manifest(raw_manifest: manifest.Manifest, subdir: str, path: str = '') -> Manifest:
+    # This cast is a bit of a hack to deal with proc-macro
+    lib = _fixup_raw_mappings(raw_manifest.get('lib', {}))
+
+    # We need to set the name field if it's not set manually,
+    # including if other fields are set in the lib section
+    lib.setdefault('name', raw_manifest['package']['name'])
+
+    return Manifest(
+        Package(**_fixup_raw_mappings(raw_manifest['package'])),
+        {k: Dependency.from_raw(v) for k, v in raw_manifest.get('dependencies', {}).items()},
+        {k: Dependency.from_raw(v) for k, v in raw_manifest.get('dev-dependencies', {}).items()},
+        {k: Dependency.from_raw(v) for k, v in raw_manifest.get('build-dependencies', {}).items()},
+        Library(**lib),
+        [Binary(**_fixup_raw_mappings(b)) for b in raw_manifest.get('bin', {})],
+        [Test(**_fixup_raw_mappings(b)) for b in raw_manifest.get('test', {})],
+        [Benchmark(**_fixup_raw_mappings(b)) for b in raw_manifest.get('bench', {})],
+        [Example(**_fixup_raw_mappings(b)) for b in raw_manifest.get('example', {})],
+        raw_manifest.get('features', {}),
+        {k: {k2: Dependency.from_raw(v2) for k2, v2 in v['dependencies'].items()}
+         for k, v in raw_manifest.get('target', {}).items()},
+        subdir,
+        path,
+    )
+
+
+def _load_manifests(subdir: str) -> T.Dict[str, Manifest]:
+    filename = os.path.join(subdir, 'Cargo.toml')
+    raw = load_toml(filename)
+
+    manifests: T.Dict[str, Manifest] = {}
+
+    raw_manifest: T.Union[manifest.Manifest, manifest.VirtualManifest]
+    if 'package' in raw:
+        raw_manifest = T.cast('manifest.Manifest', raw)
+        manifest_ = _convert_manifest(raw_manifest, subdir)
+        manifests[manifest_.package.name] = manifest_
+    else:
+        raw_manifest = T.cast('manifest.VirtualManifest', raw)
+
+    if 'workspace' in raw_manifest:
+        # XXX: need to verify that python glob and cargo globbing are the
+        # same and probably write  a glob implementation. Blarg
+
+        # We need to chdir here to make the glob work correctly
+        pwd = os.getcwd()
+        os.chdir(subdir)
+        members: T.Iterable[str]
+        try:
+            members = itertools.chain.from_iterable(
+                glob.glob(m) for m in raw_manifest['workspace']['members'])
+        finally:
+            os.chdir(pwd)
+        if 'exclude' in raw_manifest['workspace']:
+            members = (x for x in members if x not in raw_manifest['workspace']['exclude'])
+
+        for m in members:
+            filename = os.path.join(subdir, m, 'Cargo.toml')
+            raw = load_toml(filename)
+
+            raw_manifest = T.cast('manifest.Manifest', raw)
+            man = _convert_manifest(raw_manifest, subdir, m)
+            manifests[man.package.name] = man
+
+    return manifests
+
+
+def load_all_manifests(subproject_dir: str) -> T.Dict[str, Manifest]:
```

Comment:
Why would we need to load all manifests? My intention was to have wrap files for Cargo subprojects with `method=cargo`, the same way I'm doing it for cmake there: https://github.com/mesonbuild/meson/pull/11730.

Reply:
The code that uses this isn't in the tree currently, obviously, but it's to deal with the fact that you may need the same cargo project multiple times, this allows us to load all of the manifests the first time a cargo subproject is used, then refer back to those manifests later, either because you have some kind of parallel dependency or because you have two separate rust dependencies with the same dependency, ie:

```
C   C
|   |
B   D
 \ /
  A
```
or something like:
```
B        B
|        |
A        C
```

Reply:
Obviously when we get to the point of being able to use cargo wraps then we'll need to update the approach somehow, but I think for the current solution this is fine, and it can be fixed when cargo wraps land

Reply:
I have cargo wraps working locally on top of this PR (and on top of https://github.com/mesonbuild/meson/pull/11730), that's why I noticed this. But I'm fine leaving that there for now and see later how that fits.